measureme/src/counters.rs

```diff
@@ -0,0 +1,869 @@
+                ..perf_event_attr::default()
+            };
+
+            // Only record same-thread, any CPUs, and only userspace (no kernel/hypervisor).
+            // NOTE(eddyb) `pid = 0`, despite talking about "process id", means
+            // "calling process/thread", *not* "any thread in the calling process"
+            // (i.e. "process" is interchangeable with "main thread of the process")
+            // FIXME(eddyb) introduce per-thread counters and/or use `inherit`
+            // (and `inherit_stat`? though they might not be appropriate here)
+            // to be able to read the counter on more than just the initial thread.
+            let pid = 0;
+            let cpu = -1;
+            let group_fd = -1;
+            attrs.set_exclude_kernel(1);
+            attrs.set_exclude_hv(1);
+
+            let file = unsafe {
+                let fd =
+                    perf_event_open(&mut attrs, pid, cpu, group_fd, PERF_FLAG_FD_CLOEXEC.into());
+                if fd < 0 {
+                    Err(std::io::Error::from_raw_os_error(-fd))
+                } else {
+                    Ok(fs::File::from_raw_fd(fd))
+                }
+            };
+            let file = file.map_err(|e| format!("perf_event_open failed: {:?}", e))?;
+
+            let mmap = unsafe {
+                MmapOptions::new()
+                    .len(mem::size_of::<perf_event_mmap_page>())
+                    .map(&file)
+            };
+            let mmap = mmap.map_err(|e| format!("perf_event_mmap_page: mmap failed: {:?}", e))?;
+
+            let mut counter = Counter { mmap, reg_idx: 0 };
+
+            let (version, compat_version, caps, index, pmc_width) = counter
+                .access_mmap_page_with_seqlock(|mp| {
+                    (
+                        mp.version,
+                        mp.compat_version,
+                        unsafe { mp.__bindgen_anon_1.__bindgen_anon_1 },
+                        mp.index,
+                        mp.pmc_width,
+                    )
+                });
+
+            info!(
+                "Counter::new: version={} compat_version={} index={:#x}",
+                version, compat_version, index,
+            );
+
+            if caps.cap_user_rdpmc() == 0 {
+                return Err(format!(
+                    "perf_event_mmap_page: missing cap_user_rdpmc{}",
+                    if caps.cap_bit0_is_deprecated() == 0 && caps.cap_bit0() == 1 {
+                        " (ignoring legacy/broken rdpmc support)"
+                    } else {
+                        ""
+                    }
+                )
+                .into());
+            }
+
+            if index == 0 {
+                return Err(format!(
+                    "perf_event_mmap_page: no allocated hardware register (ran out?)"
+                )
+                .into());
+            }
+            counter.reg_idx = index - 1;
+
+            if (cfg!(not(accurate_seqlock_rdpmc)) || true) && pmc_width != 48 {
```

Comment:
I think this can just be 

```suggestion
            if (cfg!(not(accurate_seqlock_rdpmc))) && pmc_width != 48 {
```

Reply:
The `|| true` are so that the `cfg`s are actually ignored, since I didn't want them to accidentally become `measureme` toggles.